[2.4.0] PyO3 PanicException in RemoveIdentityEquivalent: "Cannot clone pointer into Python heap without the thread being attached"

### Environment
- **qiskit**: 2.4.0 (regression from 2.3.1)
- **qiskit-aer**: 0.17.2
- **Python**: 3.10, 3.11, 3.12 (all affected)
- **Platforms**: ubuntu-latest and macos-latest GitHub runners (all affected)

### What's happening

With qiskit 2.4.0, `transpile()` at optimization_level >= 2 panics from PyO3 when reaching the `RemoveIdentityEquivalent` pass:

```
qiskit/transpiler/passes/optimization/remove_identity_equiv.py:74: PanicException
E   pyo3_runtime.PanicException: Cannot clone pointer into Python heap without the thread being attached.
```

Traceback:

```
qiskit/compiler/transpiler.py:298: in transpile
    out_circuits = pm.run(circuits, callback=callback, num_processes=num_processes)
qiskit/transpiler/passmanager.py:419: in run
    return super().run(circuits, output_name, callback, num_processes=num_processes)
qiskit/transpiler/passmanager.py:456: in wrapper
    return meth(*meth_args, **meth_kwargs)
qiskit/transpiler/passmanager.py:196: in run
    return super().run(...)
qiskit/passmanager/passmanager.py:235: in run
    out = [
qiskit/passmanager/passmanager.py:236: in <listcomp>
    _run_workflow(...)
qiskit/passmanager/passmanager.py:309: in _run_workflow
    passmanager_ir, final_state = flow_controller.execute(...)
qiskit/passmanager/base_tasks.py:215: in execute
    passmanager_ir, state = next_task.execute(...)
qiskit/transpiler/basepasses.py:167: in execute
    new_dag, state = super().execute(...)
qiskit/passmanager/base_tasks.py:97: in execute
    ret = self.run(passmanager_ir)

self = <qiskit.transpiler.passes.optimization.remove_identity_equiv.RemoveIdentityEquivalent ...>
dag = <qiskit._accelerate.circuit.DAGCircuit object at 0x...>

    def run(self, dag: DAGCircuit) -> DAGCircuit:
>       remove_identity_equiv(dag, self._approximation_degree, self._target)
E       pyo3_runtime.PanicException: Cannot clone pointer into Python heap without the thread being attached.
```

The panic originates inside the Rust-side `remove_identity_equiv` function, not on a multiprocessing path — `num_processes` is being threaded through but the failing execution is single-threaded when we observe this.

### Regression window

- 2.3.1 (2026-03-12): green ✅
- 2.4.0 (2026-04-16 15:46 UTC): panic ❌

Since 2.4.0 is the first stable release to run on PyO3 0.28 (PRs #15670, #15939), this looks like a case where the Rust code holds or clones a `Py<T>` without a `Python::attach` guard — a contract PyO3 0.28 tightened.

### Reproduction

We hit this across our test suite whenever `vqe.compile()` calls `transpile(circuit, basis_gates=[...], optimization_level=3)` on circuits composed from `qiskit.circuit.library.efficient_su2` / `excitation_preserving` / similar n-local ansätze. A minimal synthetic repro (untested against 2.4.0 from my end — we've pinned to 2.3.1 — but fits the call path):

\`\`\`python
from qiskit import QuantumCircuit
from qiskit.compiler import transpile
from qiskit.circuit.library import efficient_su2

qc = QuantumCircuit(2)
qc.compose(efficient_su2(2, reps=1), inplace=True)
qc.assign_parameters([0.0] * qc.num_parameters, inplace=True)

transpile(qc, basis_gates=[\"rx\", \"ry\", \"rz\", \"cx\"], optimization_level=3)
\`\`\`

If that doesn't reproduce, a fuller path is: build a VQE ansatz with ~10 qubits of repeated `EfficientSU2` + `ExcitationPreserving` layers and transpile to `["rx", "ry", "rz", "cxx"]` basis at `optimization_level=3`. Happy to provide a full-repro script if useful.

### Downstream mitigation

We've pinned `qiskit<2.4` in our project until a 2.4.x patch ships.

### Observed passes

Confirmed failing pass: `qiskit.transpiler.passes.optimization.remove_identity_equiv.RemoveIdentityEquivalent`. Other passes in the preset passmanagers may have the same issue; we haven't enumerated.

### Suggested places to look
- `crates/accelerate/src/remove_identity_equiv.rs` (the callee at `remove_identity_equiv.py:74`)
- Any Rust call site that clones `Py<T>` / `Bound<T>` or traverses the DAG without an active `Python::attach`

--- a/crates/transpiler/src/passes/remove_identity_equiv.rs
+++ b/crates/transpiler/src/passes/remove_identity_equiv.rs
@@ -242,9 +242,23 @@ pub fn py_remove_identity_equiv(
     approx_degree: Option<f64>,
     target: Option<&Target>,
 ) -> PyResult<()> {
+    // TODO: This is a hack to avoid panicking in the case that the global phase contains `Py`
+    // pointers (such as backrefs to `ParameterVector` objects in an expression `Symbol`) that would
+    // get cloned when updating the global phase.  It's easier to do it out here than to try to
+    // reattach to Python-space within a pure-Rust function but only if we can spot a hiding `Py`
+    // (or we'd break standalone C).
+    //
+    // This doesn't account for control-flow blocks which _also_ might have set global phases, byt
+    // `run_remove_identity_equiv` as of Qiskit 2.4 doesn't recurse, so the hack should hold.
+    let old_phase = dag.global_phase().clone();
+    dag.set_global_phase_f64(0.0);
+
     // Explicitly release GIL because threads may call Python to get
     // the matrix for a PyGate
-    py.detach(|| run_remove_identity_equiv(dag, approx_degree, target))
+    py.detach(|| run_remove_identity_equiv(dag, approx_degree, target))?;
+
+    dag.add_global_phase(&old_phase)?;
+    Ok(())
 }
 
 pub fn run_remove_identity_equiv(

--- a/releasenotes/notes/fix-panic-remove-identity-equiv-vector-28396cfeee94a268.yaml
+++ b/releasenotes/notes/fix-panic-remove-identity-equiv-vector-28396cfeee94a268.yaml
@@ -1,0 +1,9 @@
+---
+fixes:
+  - |
+    Fixed a panic in :class:`.RemoveIdentityEquivalent` when the global phase of the circuit contained
+    :class:`.ParameterVectorElement` objects.  The message was:
+
+        Cannot clone pointer into Python heap without the thread being attached.
+
+    See `#16053 <https://github.com/Qiskit/qiskit/issues/16053>`__.

--- a/test/python/transpiler/test_remove_identity_equivalent.py
+++ b/test/python/transpiler/test_remove_identity_equivalent.py
@@ -15,7 +15,7 @@
 import ddt
 import numpy as np
 
-from qiskit.circuit import Parameter, QuantumCircuit, QuantumRegister, Gate
+from qiskit.circuit import Parameter, QuantumCircuit, QuantumRegister, Gate, ParameterVector
 from qiskit.circuit.library import (
     CPhaseGate,
     RXGate,
@@ -212,6 +212,16 @@ class TestRemoveIdentityEquivalent(QiskitTestCase):
         qc.append(GlobalPhaseGate(theta), [])
         transpiled = RemoveIdentityEquivalent()(qc)
         self.assertEqual(qc, transpiled)
+
+    def test_no_panic_on_parametervector_phase(self):
+        """Regression test for gh-16053."""
+        pv = ParameterVector("v", 1)
+        qc = QuantumCircuit(1, global_phase=pv[0])
+        qc.rz(0.0, 0)
+        qc = RemoveIdentityEquivalent()(qc)
+
+        expected = QuantumCircuit(1, global_phase=pv[0])
+        self.assertEqual(qc, expected)
 
     def test_pauli_evo_equals_stdgate(self):
         """Test the Pauli evolution gate is consistent with std gates."""

